### ISSUE TITLE
data.current(..., 'close') vs data.history(..., 'close', ...)

### ISSUE THREAD

**Issue by eranhirs** [enigmampc/catalyst#392]
Dear Catalyst Maintainers,

# Description of Issue
I get different values when running `data.current` vs `data.history`, here are the re-production steps:
1. `run_algorithm` with `bitfinex` and `quote_currency` should be `btc`
2. When `data.current_dt` is equal to `2018-04-16 00:00:00+00:00` , run `data.current(symbol('neo_btc'), 'close')`, you will get the value `0.00846018`
3. Run `data.history(symbol('neo_btc'), ['close'], 1000, '60M')`, the value for `2018-04-16 00:00:00+00:00` is now `0.0084583`.

* What did you expect to happen?
Both values to be the same. Actually, I don't know which one is the correct one.

* What happened instead?
Two different values.

## What steps have you taken to resolve this already?
I noticed that when using `.history(...)` the correct value is placed 1 hour before, specifically `2018-04-15 23:00:00+00:00`

# Anything else?

Thanks!

**Comment by lenak25** [comment]
Thanks for reporting.

Could you please provide the following  information: 

1. The mode of algorithm (I assume it is backtest, right?) and the data frequency in use 
2. The exchange you are using
3. catalyst version

Thanks,

Lena

**Comment by eranhirs** [comment]
Thanks for the fast response!
1. mode is backtesting, data frequency is minute
2. exchange is bitfnex
3. catalyst version is 0.5.14

**Comment by eranhirs** [comment]
I think we can close this.
I just now realized that I am running `.close` in a `minute` setting not an `hour` setting so it gives me the last price of `2018-04-16 00:00:00+00:00` minute, not hour

**Comment by sam31415** [comment]
Why would you expect that during a period, data.current would return the close price for this period? This would introduce an obvious look ahead bias: you cannot know the close price until after the period has ended.

**Comment by eranhirs** [comment]
@sam31415 I am not sure what is the problem.
I put the order at the end of the current period (the `close` price).
Catalyst engine then simulates the buy at the order of the next `close` price.

**Comment by sam31415** [comment]
I don't think it works like this, at least from my experience on Quantopian. As far as I know Catalyst start working at the beginning of the period, not at the end, which imho makes sense.

**Comment by eranhirs** [comment]
From debugging the code I noticed that the orders are executed 1 bar after the one you put the orders in.
Check @AvishaiW comment in this issue:
https://github.com/enigmampc/catalyst/issues/295#issuecomment-382302818

**Comment by lenak25** [comment]
Thanks @eranhirs , the difference in the price is indeed due to the different period used (minute vs hour).
Regarding the `data.current` - @eranhirs observation is correct - the current bar (which in minute frequency is a minute bar) holds the entire information for this period. 
Meaning that if `current_dt` is  2018-04-16 00:00:00+00:00 the OHLCV information is for the period 2018-04-16 00:00:00+00:00 -  2018-04-16 00:00:59+00:00. 
@sam31415 , this is not a bias, because as @eranhirs explained, all decisions are taken into consideration in the next bar. IMHO this is Zipline logic, but if you feel otherwise, please let us know.

**Comment by sam31415** [comment]
@lenak25 One thing I can tell you for sure is that in live trading, at 2018-04-16 00:00:00, current_dt is equal to 2018-04-16 00:00:00. If now you're telling me that in the backtest in minute data, when current_dt equals 2018-04-16 00:00:00, you have access to the close price at 2018-04-16 00:01:00, then there is a look ahead bias. In live trading at 2018-04-16 00:00:00, you don't know what will be the close price of the current minute bar. 

Are you really sure about what you just wrote? If yes, that's worrisome. And I really doubt such an obvious flaw would be present in Zipline. I'll check the source when I get some time.

**Comment by lenak25** [comment]
@sam31415 that is a good point and I agree with you. 
We will investigate this further to understand where this behaviour came from.
Thanks for bringing this up!

**Comment by sam31415** [comment]
@STOpandthink I assume this is current_dt during backtesting, right? Then it means that there is indeed an inconsistency between the backtest mode and the live mode, because in live trading, current_dt is really the current time.

**Comment by lenak25** [comment]
Thanks everyone for your inputs, we have investigated this issue and made the required adjustments, included in release 0.5.19.

The problem was due to the different definitions of the OHLCV intervals in the cryptocurrency exchanges versus the stock market.

Cryptocurrency trading exchanges use left-bounded intervals when defining OHLCV candles:
```
9:31 interval: 9:31:00 am <= t < 9:32:00 am
9:32 interval: 9:32:00 am <= t < 9:33:00 am
9:33 interval: 9:33:00 am <= t < 9:34:00 am
```

To meet the different nature of the crypto data definition versus the stock market data, Catalyst code had to be altered as follows: when the minute clock ticks at the beginning of every minute `data.current` returns the closing price from the previous minute and `data.history` last bar matches `data.current` providing the price up to the minute prior to the current one.
Meaning that if `current_dt == 2018-04-16 00:01:00` the `data.current` will return the  information of the `2018-04-16 00:00:00 - 2018-04-16 00:00:59` period.

A more [detailed explanation](https://enigma.co/catalyst/intervals.html) was added to our documentation as well.

@sam31415 @eranhirs @STOpandthink, thanks again for reporting!
